Implement the hot wallet for comakery platform

TheMMaciek opened the pull request.

TheMMaciek committed:
Implement the hot wallet for comakery platform

marcinwadon commented on src/main/resources/application.conf:
```diff
@@ -0,0 +1,15 @@
+    load-balancer {
+        host = ""
+        port = 
+        }
```
indent

marcinwadon commented on src/main/scala/org.constellation.cmhotwallet/App.scala:
```diff
@@ -0,0 +1,118 @@
+    for {
+      params <- updatedParams
+      _ <- keyTool.generateAndStoreKeyPair(params.cmKeystore, params.cmAlias, params.cmStorepass, params.cmKeypass)
+      _ <- println(s"${GREEN}KeyPair created successfully!${RESET}").pure[F].attemptT
```
can you use slf4j logger instead of println? you can log to console as well

marcinwadon commented on src/main/scala/org.constellation.cmhotwallet/model/request/CLTransaction.scala:
```diff
@@ -0,0 +1,49 @@
+    TypedEdgeHash
+  }
+
+  implicit val clLastTransactionRefEncoder: Encoder[LastTransactionRef] = deriveEncoder
+  implicit val clLastTransactionRefDecoder: Decoder[LastTransactionRef] = deriveDecoder
```
Why it's needed there? I believe it's specified in the `org.constellation.wallet` as well

TheMMaciek replied:
The wallet itself didn't send the transaction so it didn't need the encoders for http client - it used circe auto derivation to export the transaction to file and read from file.

marcinwadon commented on src/main/scala/org.constellation.cmhotwallet/CoMakerClient.scala:
```diff
@@ -0,0 +1,135 @@
+class CoMakerClient(keyTool: Ed25519KeyTool) {
+  import CoMakerClient._
+
+  private val secureRandom: SecureRandom = try {
+    SecureRandom.getInstance("NativePRNGNonBlocking")
+  } catch {
+    case _: Throwable => SecureRandom.getInstanceStrong
+  }
+
+  private def nonce: String = BigInt(120, secureRandom).toString(16)
+
+  private def timestamp: String = Instant.now.getEpochSecond.toString
+
+  private def getUri(path: String)(config: CoMakeryConfig): Uri =
+    Uri(scheme = Some(Scheme.http), authority = Some(Authority(host = RegName(config.host), port = config.port.some)))
+      .addPath(config.apiSuffix)
+      .addPath(path)
```
Can you make it in more functional way? The rest of code is written in such manner so it would be good to keep it unified over the codebase.

TheMMaciek replied:
It looks like a builder, but actually underneath it uses .copy, so it's functional in the end.

marcinwadon commented on build.sbt:
```diff
@@ -0,0 +1,46 @@
+  "org.bouncycastle"         %  "bcprov-jdk15on"      % versions.bouncycastle,
+  "org.bouncycastle"         %  "bcpkix-jdk15on"      % versions.bouncycastle,
+  "com.github.pureconfig"    %% "pureconfig"          % versions.pureconfig,
+  "org.scalatest"            %% "scalatest"           % versions.scalaTest % "test",
```
It would look better if you will specify test dependencies separately:

```
testDependencies = Seq(
 ...
).map(_ % "test")

libraryDependencies ++= Seq(...) ++ testDependencies
```

marcinwadon commented on assembly.sh:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+sbt assembly
```
It was a leftover in the main project. I believe it's not needed there

TheMMaciek committed:
Make port and transfer id optional and use https connection

TheMMaciek committed:
Setup logging and use it instead of printing to console

TheMMaciek committed:
Add confirmation step before sending the transaction

TheMMaciek committed:
Change authentication to use Api-Transaction-Key

TheMMaciek committed:
Handle case when there are not transfers to pay

TheMMaciek committed:
Remove not used functionality and clean up the code

TheMMaciek committed:
Remove not used dependencies

TheMMaciek committed:
Remove assembly.sh

marcinwadon commented on src/main/scala/org.constellation.cmhotwallet/App.scala:
```diff
@@ -0,0 +1,96 @@
+
+  def payTransfer[F[_]: Logger](config: Config, projectConfig: ProjectConfig, cliParams: CliConfig)(
+    client: Resource[F, Client[F]]
+  )(implicit F: Sync[F], C: ConcurrentEffect[F]) =
```
Why you need both? ConcurrentEffect is a Sync as well

TheMMaciek committed:
Fix decoding the constellation transaction hash

By decoding as String the hash was wrapped in quotation marks, which propageted as part of a hash causing issues with not found transaction on block-explorer

TheMMaciek committed:
Remove not needed doubled effect types

TheMMaciek committed:
Improve logging transaction processing steps

TheMMaciek committed:
Adjust application configuration to production urls
